File: backend/app/schemas/invitation.py
"""Invitation schemas for API requests and responses."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class InvitationCreate(BaseModel):
    """Schema for creating a single invitation."""

    email: EmailStr
    first_name: str | None = Field(None, max_length=100)
    last_name: str | None = Field(None, max_length=100)
    list_type: str = Field(
        default="standard",
        pattern="^(standard|list_1000|list_2000)$",
    )


class InvitationResponse(BaseModel):
    """Response schema for invitation data."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    email: EmailStr
    first_name: str | None = None
    last_name: str | None = None
    status: str
    created_at: datetime
    expires_at: datetime | None = None  # None for activated users
    used_at: datetime | None = None  # Activation date for activated users


class BulkInvitationResult(BaseModel):
    """Response schema for bulk invitation creation."""

    total: int
    created: int
    duplicates: int
    errors: list[dict]


class InvitationResendResponse(BaseModel):
    """Response schema for invitation resend."""

    id: str
    email: EmailStr
    status: str
    expires_at: datetime
    message: str = "Invitation resent successfully"


class BulkDeleteRequest(BaseModel):
    """Request schema for bulk invitation deletion."""

    ids: list[str] = Field(..., min_length=1, max_length=100)


class BulkDeleteResult(BaseModel):
    """Response schema for bulk invitation deletion."""

    total: int
    deleted: int
    not_found: int

File: frontend/src/types/invitation.ts
/**
 * Invitation types for API requests and responses.
 */

export type InvitationStatus = 'pending' | 'sent' | 'activated' | 'expired' | 'cancelled';

export type ListType = 'standard' | 'list_1000' | 'list_2000';

export interface Invitation {
  id: string;
  email: string;
  firstName: string | null;
  lastName: string | null;
  status: InvitationStatus;
  createdAt: string;
  expiresAt: string | null; // null for activated users
  usedAt: string | null; // activation date for activated users
}

export interface InvitationCreateRequest {
  email: string;
  firstName?: string | null;
  lastName?: string | null;
  listType?: ListType;
}

export interface InvitationCreateResponse extends Invitation {}

export interface BulkInvitationResult {
  total: number;
  created: number;
  duplicates: number;
  errors: BulkInvitationError[];
}

export interface BulkInvitationError {
  email: string;
  error: string;
}

export interface InvitationResendResponse {
  id: string;
  email: string;
  status: string;
  expiresAt: string;
  message: string;
}

export type InvitationStatusFilter = 'pending' | 'expired' | 'activated';

export interface BulkDeleteRequest {
  ids: string[];
}

export interface BulkDeleteResult {
  total: number;
  deleted: number;
  notFound: number;
}

export interface BulkResendResult {
  total: number;
  resent: number;
  skipped: number;
}

export interface DepositorLookup {
  found: boolean;
  firstName?: string;
  lastName?: string;
  participationCount?: number;
  lastEditionName?: string;
  preferredListType?: string;
}
